src/Service/MagicLinks/Validation/MagicLinkTemplateUrl.cs

```diff
@@ -0,0 +1,25 @@
+using System.ComponentModel.DataAnnotations;
+
+namespace Passwordless.Service.MagicLinks.Validation;
+
+[AttributeUsage(AttributeTargets.Property | AttributeTargets.Field)]
+public class MagicLinkTemplateUrl : ValidationAttribute
+{
+    private const string TokenTemplate = "<token>";
+
+    protected override ValidationResult IsValid(object value, ValidationContext validationContext)
+    {
+        if (value is not string stringValue || string.IsNullOrWhiteSpace(stringValue))
+            return new ValidationResult("Value must be a non-empty string.");
+
+        if (!stringValue.Contains(TokenTemplate))
+            return new ValidationResult("Value must contain the `<token>` template.");
```

Comment:
Since we already have the constant defined...

```suggestion
            return new ValidationResult($"Value must contain the `{TokenTemplate}` template.");
```